src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+                }
+            }
+            catch (Exception)
+            {
+                PeerNotReachable?.Invoke(this, peer);
+                throw;
+            }
+            finally
+            {
+                _pendingConnections.TryRemove(peer, out _);
+            }
+        }
+
+        /// <summary>
+        ///   Create a stream to the peer that talks the specified protocol.
+        /// </summary>
+        /// <param name="peer">
+        ///   The remote peer.
+        /// </param>
+        /// <param name="protocol">
+        ///   The protocol name, such as "/foo/0.42.0".
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is the new <see cref="Stream"/> to the <paramref name="peer"/>.
+        /// </returns>
+        /// <remarks>
+        ///   <para>
+        ///   When finished, the caller must <see cref="Stream.Dispose()"/> the
+        ///   new stream.
+        ///   </para>
+        /// </remarks>
+        public async Task<Stream> DialAsync(Peer peer,
+            string protocol,
+            CancellationToken cancel = default)
+        {
+            peer = RegisterPeer(peer);
+
+            // Get a connection and then a muxer to the peer.
+            var connection = await ConnectAsync(peer, cancel).ConfigureAwait(false);
+            var muxer = await connection.MuxerEstablished.Task.ConfigureAwait(false);
+
+            // Create a new stream for the peer protocol.
+            var stream = await muxer.CreateStreamAsync(protocol, cancel).ConfigureAwait(false);
+            try
+            {
+                await connection.EstablishProtocolAsync("/multistream/", stream, cancel).ConfigureAwait(false);
+
+                await Message.WriteAsync(protocol, stream, cancel).ConfigureAwait(false);
+                var result = await Message.ReadStringAsync(stream, cancel).ConfigureAwait(false);
+                if (result != protocol) throw new Exception($"Protocol '{protocol}' not supported by '{peer}'.");
+
+                return stream;
+            }
+            catch (Exception)
+            {
+                stream?.DisposeAsync();
+                throw;
+            }
+        }
+
+        /// <summary>
+        ///   Establish a duplex stream between the local and remote peer.
+        /// </summary>
+        /// <param name="remote"></param>
+        /// <param name="addrs"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        private async Task<PeerConnection> DialAsync(Peer remote,
+            IEnumerable<MultiAddress> addrs,
+            CancellationToken cancel)
+        {
+            _log.Debug($"Dialing {remote}");
+
+            if (remote == LocalPeer) throw new Exception("Cannot dial self.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)